### PR TITLE
Improvements to learned round

### PR DESCRIPTION
**Pull request by Giuseppe5**
## Reason for this PR

Fix entrypoint for learned scale
Fix training with float32 + amp

## Testing Summary

NA

## Risk Highlight

- [ ] This PR includes code from another work (please detail).
- [ ] This PR contains API-breaking changes.
- [ ] This PR depends on work in another PR (please provide links/details).
- [ ] This PR introduces new dependencies (please detail).
- [ ] There are coverage gaps not covered by tests.
- [ ] Documentation updates required in subsequent PR.

## Checklist

- [ ] Code comments added to any hard-to-understand areas, if applicable.
- [ ] Changes generate no new warnings.
- [ ] Updated any relevant tests, if applicable.
- [ ] No conflicts with destination `dev` branch.
- [ ] I reviewed my own code changes.
- [ ] Initial CI/CD passing.
- [ ] 1+ reviews given, and any review issues addressed and approved.
- [ ] Post-review full CI/CD passing.

### CODE REVIEW

File: src/brevitas_examples/llm/llm_quant/learned_round_utils.py [hunk 1]
```diff
@@ -114,23 +114,23 @@ def apply_learned_round(
...
     use_best_model: bool = True,
-    use_amp: bool = True,
     amp_dtype: torch.dtype = torch.float16,
     loss_scaling_factor: float = 1000,
     lr_scheduler: Optional[str] = "linear",
     optimizer_kwargs: Optional[Dict] = None,
     lr_scheduler_kwargs: Optional[Dict] = None,
     learned_round_loss_kwargs: Optional[Dict] = None,
+    scale_optimizer_class: Optional[str] = None,
+    scale_optimizer_kwargs: Optional[dict] = None,
```
Comment (pablomlago) [inline src/brevitas_examples/llm/llm_quant/learned_round_utils.py:127]:
```suggestion
    scale_optimizer_kwargs: Optional[Dict] = None,
```

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 2]
```diff
@@ -508,25 +510,32 @@ def _optimize_learned_round_block(
         init_loss = -1.0
         last_best_iter = self.iters
 
+        scaler = None
+        use_amp = next(block.parameters()).dtype == torch.float32
```
Comment (pablomlago) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:514]: Why torch.float32? If the LLM dtype is torch.float16, this would be False and no gradient scaling would be used.

Comment (Giuseppe5) [reply src/brevitas_examples/common/learned_round/learned_round_optimizer.py:514]: Gradient scaling and AMP are to be used only with higher precision data type (float32) and take advantage of fp16 in selected layers.

If the network is already cast to float16, we don't need mixed precision, and we do manual scaling of the loss later

Comment (Giuseppe5) [reply src/brevitas_examples/common/learned_round/learned_round_optimizer.py:514]: GradScaler specifically raise an error if the original network is fully cast to float16

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 3]
```diff
@@ -492,12 +496,10 @@ def _optimize_learned_round_block(
...
             optimizer_scale = self.scale_optimizer_class(
                 scale_params,
-                lr=self.optimizer_scale_lr,
-                **self.optimizer_kwargs,
+                **self.scale_optimizer_kwargs,
             )
             lr_scheduler_scale = (
-                self.lr_scheduler_class(
-                    optimizer_scale, start_factor=1, end_factor=0, total_iters=600)
+                self.lr_scheduler_class(optimizer_scale, **self.lr_scheduler_kwargs)
```
Comment (pablomlago) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:502]: Should we allow for different LRSchedulers for the scale/learned round parameters?

File: src/brevitas_examples/common/learned_round/learned_round_optimizer.py [hunk 4]
```diff
@@ -508,25 +510,32 @@ def _optimize_learned_round_block(
...
                               dtype=self.amp_dtype):
+                    # Run block forward to obtain quant outputs
+                    quant_outs = block_forward(block, inputs)
+                    fp_outs = send_to_device(fp_outs, quant_outs.device)
                     loss, loss_components = block_loss(quant_outs, fp_outs)
             else:
+                # Run block forward to obtain quant outputs
+                quant_outs = block_forward(block, inputs)
+                fp_outs = send_to_device(fp_outs, quant_outs.device)
                 loss, loss_components = block_loss(quant_outs.to(torch.float32), fp_outs.to(torch.float32))
```
Comment (pablomlago) [inline src/brevitas_examples/common/learned_round/learned_round_optimizer.py:539]: The code for each condition is almost exactly the same. Maybe we could have `autocast(enabled=use_amp, ...)` and just have a conditional for upcasting the outputs to float32 before computing the loss to avoid repeting block_forward/send_to_device/block_loss.

Comment (Giuseppe5) [reply src/brevitas_examples/common/learned_round/learned_round_optimizer.py:539]: I understand it's less code, but I think it will be more confusing. I am leaving as it is. Extra verbosity for clarity, I am happy with that.
